test/e2e/e2e_test.go

```diff
@@ -511,6 +518,21 @@ func pollForCARotation(t *testing.T, client *kubernetes.Clientset, caCertPEM, ca
 	return secret
 }
 
+// pollForCARecreation polls for the signing secret to be re-created in
+// response to CA secret deletion.
+func pollForCARecreation(client *kubernetes.Clientset) error {
+	return wait.PollImmediate(time.Second, rotationPollTimeout, func() (bool, error) {
+		_, err := client.CoreV1().Secrets(serviceCAControllerNamespace).Get(context.TODO(), signingKeySecretName, metav1.GetOptions{})
+		if err != nil && errors.IsNotFound(err) {
+			return false, nil
+		}
+		if err != nil {
+			return false, err
```

Comment:
(No action required) Classic polling mistake - returning an error exits rather than triggering retry as intended. No blame intended - lots of examples of this in this file. Best practice for handling potentially recoverable failure in a polling function is to log the error and not return it.

Reply:
Thanks, I was treating the server as “stable/available by assumption” and I didn’t realize that’s a bad assumption for HA code.

I’m not going to change this one example as an exception, but if you’d prefer me to change all of them, I can do that.